src/loaders/workspace_uploader/workspace_uploader.py

```diff
@@ -0,0 +1,571 @@
+        loader_helper.create_hardlink_between_files(dest_file, src_file)
+
+    return data_dir
+
+
+def _create_entries_in_sourcedata_workspace(
+        upload_env_key: str,
+        workspace_id: int,
+        assembly_names: list[str],
+        assembly_name_to_dir: dict[str, str],
+        assembly_name_to_upa: dict[str, str],
+        output_dir: str,
+) -> list[str]:
+    """
+    Create a standard entry in sourcedata/workspace for each assembly.
+    Hardlink to the original assembly file in sourcedata to avoid duplicating the file.
+    Update the uploaded.yaml file in the genome directory with assembly names and upa info.
+    """
+    upas = list()
+    for assembly_name in assembly_names:
+        try:
+            assembly_dir = assembly_name_to_dir[assembly_name]
+        except KeyError as e:
+            raise ValueError(f"Unable to find assembly {assembly_name}") from e
+
+        src_file = _get_source_file(assembly_dir, assembly_name)
+
+        try:
+            upa = assembly_name_to_upa[assembly_name]
+        except KeyError as e:
+            raise ValueError(f"Unable to find assembly {assembly_name} from target workspace") from e
+        
+        upa_dir = os.path.join(output_dir, upa)
+        os.makedirs(upa_dir, exist_ok=True)
+        upas.append(upa)
+
+        dest_file = os.path.join(upa_dir, assembly_name)
+        loader_helper.create_hardlink_between_files(dest_file, src_file)
+        _update_upload_status_yaml_file(upload_env_key, workspace_id, upa, assembly_dir, assembly_name)
+
+    return upas
+
+
+def process_input(conf: Conf) -> None:
+    """
+    Process input from input_queue and put the result in output_queue.
+    """
+    while True:
+        task = conf.input_queue.get(block=True)
+        if not task:
+            print("Stopping")
+            break
+
+        upa = None
+        workspace_name, assembly_path, assembly_name = task
+        try:
+            upa = _upload_assembly_to_workspace(conf, workspace_name, assembly_path, assembly_name)
+        except Exception as e:
+            print(e)
+            print(f"Failed assembly name: {assembly_name}")
+        conf.output_queue.put((assembly_name, upa))
+
+
+def upload_assembly_files_in_parallel(
+        conf: Conf,
+        workspace_name: str,
+        assembly_files: list[str],
+) -> Tuple[dict[str, str], list[str]]:
+    """
+    Upload assembly files to the target workspace in parallel using multiprocessing
+
+    conf: Conf object
+    workspace_name: target workspace name
+    assembly_files: list of assembly files to upload
+    """
+    assembly_files_len = len(assembly_files)
+    print(f"Start uploading {assembly_files_len} assembly files with {conf.workers} workers\n")
+
+    # Put the assembly files in the input_queue
+    counter = 1
+    for assembly_name in assembly_files:
+
+        if counter % 5000 == 0:
+            print(f"{round(counter / len(assembly_files), 4) * 100}% finished at {datetime.now()}")
```

Comment:
Updated! (Also add % how many assemblies have been processed)

Reply:
The text for the first counter is now accurate, but the behavior in _process_input is misleading, since there's no guarantee the tasks will be processed in counter order. If one worker happens to be processing a lot of really small genomes and has an task with a counter value of 6000 in the queue, and another worker has to deal with many large genomes but has a task with a 3000 counter value, the "6000 genomes done" message could be printed out before the "3000 genomes done" message.

Also if I had to chose between a X / Y print out or a percentage only I'd rather have the former since you can calculate the percentage from that, but you can't go in the reverse direction. Or have both

Reply:
Updated, now we have both X/Y and percentage. See the following examples:

Jobs added to the queue: 4/10, Percentage: 40.00%, Time: 2023-08-14 16:02:42.605139

Assemblies processed: 6/10, Percentage: 60.00%, Time: 2023-08-14 16:03:34.204360

Reply:
:+1: